content/blog/2019-11-13-llvm-profiling/index.md

```diff
@@ -0,0 +1,336 @@
+
+1. Basic execution information including the number of LLVM instructions, basic blocks etc.
+2. Expensive instruction information like the number of memory operations, multiplications and branches since they are most likely to affect execution time.
+3. Function call information: the number of times each function is called.
+
+
+
+## Design Overview
+
+The LLVM architecture looks like this:
+
+<img src="llvm.jpg" style="width: 90%">
+
+
+
+IR represents intermediate language. We only need to add one pass that changes the existing IR to another IR (which includes the desired properties like profiling information), following instructions to add an LLVM pass in our instructor, [Adrian](<https://www.cs.cornell.edu/~asampson/>)'s [blog post](<https://www.cs.cornell.edu/~asampson/blog/llvm.html>). We also took help from the github repo [atomicCounter](<https://github.com/pranith/AtomicCounter/blob/master/AtomicCountPass/AtomicCount.cpp>) where it counts the number of all atomic instructions in the program atomically!
+
+We start from [ModulePass](<http://llvm.org/docs/WritingAnLLVMPass.html#the-modulepass-class>) and then access functions inside a Module, Basic Blocks per function and instructions per Basic Block to obtain all information. At the beginning and ending of each module, we also call custom method `initialize` to create global variables in IR  and `finalize` to print global variables. The structure of our atomic counting pass looks like this:
+
+```c++
+  struct SkeletonPass : public ModulePass {
+    static char ID;
+    SkeletonPass() : ModulePass(ID) {}
+      
+    virtual bool runOnModule(Module &M); //when there is a Module
+    virtual bool runOnFunction(Function &F, Module &M); //called by runOnModule
+    virtual bool runOnBasicBlock(BasicBlock &BB, Module &M); // called by runOnFunction
+      
+    bool initialize(Module &M); //create global variable
+    bool finialize(Module &M); //print global variable
+      
+    void createInstr(BasicBlock &bb, Constant *counter_ptr, int num);
+      
+    vector<string> atomicCounter; //keep global variable names for profiling. e.g. instr counter
+  };
+```
+
+
+
+## Implementation
+
+### Basic Execution Information
+
+We first create global variable `llvmInstrAtomicCounter` and `basicBlockAtomicCounter` with [GlobalVariable](<https://llvm.org/doxygen/classllvm_1_1GlobalVariable.html#a3ef813d6bda7e49e31cb6bf239c4e264>) constructor.
+
+```c++
+new GlobalVariable(M, I64Ty, false, GlobalValue::CommonLinkage, ConstantInt::get(I64Ty, 0), atomicCounter[i]);
+```
+
+ Then in each basic block we obtain pointer to the global variable names with [getOrInsertGlobal](<https://llvm.org/doxygen/classllvm_1_1Module.html#abd8f7242df6ecb10f429c4d39403c334>) method. After getting the instruction number in each block, we create atomic addition with [AtomicRMWInst](<https://llvm.org/doxygen/classllvm_1_1AtomicRMWInst.html#abf7e0649c7f272cc49165e579be010a5>) constructor. The complete code looks like this:
```

Comment:
It doesn't seem necessary to re-run "getOrInstertGlobal" while iterating over each basic block… you can keep a reference to it and reuse that.

Reply:
I feel I understand your question on `vector<string> atomicCounter;` now! Yes, we could have done that. Thanks for your advice.